_posts/2024-04-28-changing-loggers-level-from-unleash.adoc

```diff
@@ -0,0 +1,562 @@
+        } else {
+            return hostName.equals(logConfig.hostName);
+        }
+    }
+
+    // Unchanged.
+    private void setLoggerLevel(String loggerName, String levelName) {
+        Logger logger = Logger.getLogger(loggerName);
+        Level currentLevel = logger.getLevel();
+        Level newLevel = Level.parse(levelName);
+        if (!newLevel.equals(currentLevel)) {
+            logger.setLevel(newLevel);
+        }
+    }
+}
+----
+<1> In OpenShift, we're passing the generated pod name through the `HOST_NAME` environment variable.
+<2> That's new!
+<3> This block is used to filter hosts based on a host name prefix. That's enough for our use case, but a regular expression could be used for finer filtering.
+
+Here's how the variant payload may look like after these changes:
+
+[source,json]
+----
+[
+  {
+    "hostName": "unstable-service-7dbbcb4cc-9d9hl",
+    "category": "io.quarkus.arc",
+    "level": "FINE"
+  },
+  {
+    "hostName": "awesome-app*",
+    "category": "org.acme.SomeService",
+    "level": "WARNING"
+  },
+  {
+    "category": "org.apache.kafka.clients",
+    "level": "FINER"
+  }
+]
+----
+
+In that payload:
+
+- the first entry will affect a specific host: `unstable-service-7dbbcb4cc-9d9hl`
+- the second entry will affect all hosts whose name starts with `awesome-app`
+- the third entry will affect all hosts regardless of their names
+
+== Reverting changes automatically
+
+Changing the level of loggers through an Unleash variant should be a temporary action, mostly for troubleshooting purposes.
+This means we need to revert the level of the loggers eventually when the troubleshooting is over.
+Doing so by hand would be painful, so let's see how we can automate that.
+
+[source,java]
+----
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.getunleash.Unleash;
+import io.getunleash.Variant;
+import io.getunleash.event.UnleashSubscriber;
+import io.getunleash.repository.FeatureToggleResponse;
+import io.getunleash.variant.Payload;
+import io.quarkus.logging.Log;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static io.getunleash.repository.FeatureToggleResponse.Status.CHANGED;
+import static java.util.stream.Collectors.toSet;
+
+@ApplicationScoped
+public class LogLevelManager implements UnleashSubscriber {
+
+    private static final String UNLEASH_TOGGLE_NAME = "my-app.log-levels";
+
+    @ConfigProperty(name = "host-name", defaultValue = "localhost")
+    String hostName;
+
+    @Inject
+    Unleash unleash;
+
+    @Inject
+    ObjectMapper objectMapper;
+
+    private final Map<String, Level> originalLoggerLevels = new HashMap<>();
```

Comment:
I wonder if the Map shouldn't be made concurrent to be extra sure. I reckon it's highly unlikely you would get concurrency issues but I think it would be safer.

Reply:
Concurrency issues are very unlikely as the Unleash client fetches the toggles with an internal scheduled job every 10 seconds by default (this is configurable).

Reply:
I just replaced the `HashMap` with a `ConcurrentHashMap`. Also update the post date.